[elixir] Add missing Alchemist commands and key bindings

swaroopch opened the pull request:
Commands and key bindings found via `alchemist-refcard` command.

I've tried to adhere to Spacemacs conventions but where I couldn't find relevant
conventions, I followed Alchemist's key bindings.

syl20bnr commented on layers/+lang/elixir/packages.el:
```diff
@@ -92,10 +94,13 @@
-      "tf" 'alchemist-test-file
-      "tn" 'alchemist-test-jump-to-next-test
-      "tp" 'alchemist-test-jump-to-previous-test
+      "tT" 'alchemist-project-run-tests-for-current-file
```
What is the difference with `alchemist-mix-test-file` ?

swaroopch replied:
`alchemist-mix-test-file` will run tests for this file - if it cannot find the tests, it will ask to choose which test file to run.

`alchemist-project-run-tests-for-current-file` will run tests for this file - if it cannot find the tests, it will ask whether to create it (y/n) and then create the corresponding file and open it in a new window for writing.

syl20bnr commented on layers/+lang/elixir/packages.el:
```diff
@@ -105,7 +110,28 @@
+      "oi" 'alchemist-macroexpand-once-region
+      "oI" 'alchemist-macroexpand-once-print-region
+      "or" 'alchemist-macroexpand-region
+      "oR" 'alchemist-macroexpand-print-region)
```
I don't how macroexpand works for elixir but do you think this could be similar to `macrostep` for elisp ?
https://github.com/syl20bnr/spacemacs/blob/develop/layers/+lang/emacs-lisp/packages.el#L164-L175

swaroopch replied:
My understanding is that the elixir macroexpand is not interactive, it is meant to be "expand once" or "expand completely" for different region sizes and with optional "printing" (appending below the current region as comments).

e.g. : 
<img width="659" alt="screen shot 2017-04-20 at 5 05 41 am" src="https://cloud.githubusercontent.com/assets/42988/25230229/143e823a-2588-11e7-858c-5d03cd66f607.png">

swaroopch commented:
Also, do we really need the maintenance of `alchemist-refcard.tex`? I'm not sure why that is needed since we already have a README, it would be great to not have to update documentation twice :)

swaroopch commented on layers/+lang/elixir/packages.el:
```diff
@@ -91,10 +93,13 @@
       "tb" 'alchemist-mix-test-this-buffer
       "tt" 'alchemist-mix-test-at-point
-      "tf" 'alchemist-mix-test-file
+      "tf" 'alchemist-project-run-tests-for-current-file
```
Please note this breaking change... my goal was to make the key bindings more intuitive.

syl20bnr commented:
> Also, do we really need the maintenance of alchemist-refcard.tex? I'm not sure why that is needed since we already have a README, it would be great to not have to update documentation twice :)

It was a nice addition but you are right, we can delete them as they don't fit with the standard of Spacemacs so it feels weird to have this for only language.
PR welcome to remove them.

swaroopch commented:
@syl20bnr I have updated the PR to remove the `alchemist-refcard.{tex,pdf}` files.

I have been using this patch successfully for a few weeks now and have no issues so far.

Please let me know if any other changes are needed. 

Thanks!

JAremko commented on layers/+lang/elixir/README.org:
```diff
@@ -201,10 +206,39 @@ You find and overview of all the key-bindings on the [[https://github.com/tonini
 | ~SPC m x :~ | Run a custom execute command with =elixir= |
 | ~SPC m x b~ | Run the current buffer through =elixir=    |
 | ~SPC m x f~ | Run =elixir= with the given filename       |
-
```
One new line after a table is required :shipit:

swaroopch commented:
@TheBB I have squashed the branch.

@JAremko I have fixed the newline but now the build seems to complain about the ToC... I believe the ToC change was done automatically, not sure what I should be doing here, I'm on the latest `develop` branch, can you please guide me here?

JAremko commented:
@swaroopch  `toc-org` was updated recently https://github.com/snosov1/toc-org/commit/421956e9cc158e645e7cf88c47caff40099c0e5c  but CI uses the old version.  @TheBB  I think we need to merge this https://github.com/syl20bnr/spacemacs/pull/8931  while I'm working on redesigning `spacefmt`

JAremko commented:
@swaroopch this PR has conflict :cry:

swaroopch commented:
@JAremko Rebased.

TheBB commented:
Thanks! Cherry-picked in develop. You can safely delete your branch.
